data/lua/carryover_gold.lua

```diff
@@ -0,0 +1,248 @@
+		end
+	end
+	return res
+end
+
+--- Gets the number of villages on the map.
+---@return integer
+function carryover.get_num_villages()
+	return #(wesnoth.map.find{ gives_income = true })
+end
+
+--- Like tostring(num) but uses a prettier minus sign, to be used in the UI.
+---@return string
+function carryover.half_signed_value(num)
+	if num < 0 then
+		return "-" .. tostring(math.abs(num))
+	else
+		return tostring(num)
+	end
+	return tostring(num)
+end
+
+--- Calculates the amount of carryover gold.
+---@param side side
+---@param turns_left integer
+---@return number (finishing_bonus), number (finishing_bonus_per_turn)
+function carryover.calculate_finishing_bonus(side, turns_left)
+	local num_villages = carryover.get_num_villages()
+
+	local finishing_bonus_per_turn = num_villages * side.village_gold + side.base_income
+	local finishing_bonus = math.ceil(side.carryover_bonus * finishing_bonus_per_turn * turns_left)
+	return finishing_bonus, finishing_bonus_per_turn
+end
+
+---@return boolean
+function carryover.get_report_visible()
+	local is_observer = carryover.get_is_observer()
+	local is_replay = wesnoth.current.is_replaying
+	--return true
+	return wesnoth.scenario.end_level_data.carryover_report and (not is_observer) and (not is_replay)
+end
+
+--- Whether to show the sides names in the scenario end report.
+---@return boolean
+function carryover.get_show_side_names()
+	local num_persistent_teams = carryover.get_num_persistent_teams()
+	return num_persistent_teams > 1
+end
+
+
+--- returns a stub to show in the scenario end dialog.
+---@return string (title), string (report)
+function carryover.get_popup_text_basic()
+	local is_observer = carryover.get_is_observer()
+	local is_victory = wesnoth.scenario.end_level_data.is_victory
+
+	local report = ""
+	local title = ""
+
+	if is_observer then
+		title = _("Scenario Report")
+	elseif is_victory then
+		title = _("Victory")
+		report = report .. "<b>" .. _("You have emerged victorious!") .. "</b>"
+	else
+		title = _("Defeat")
+		report = report .. _("You have been defeated!")
+	end
+	return title, report
+end
+
+-- returns the first part of the carryover mesage, explain remaining gold
+---@param side side
+---@param info table
+---@return string
+function carryover.remaining_gold_message(side, info)
+	return "<small>\n" .. _("Remaining gold: ") .. carryover.half_signed_value(side.gold) .. "</small>"
+end
+
+---@param side side
+---@param info table
+---@return string
+function carryover.bonus_gold_message(side, info)
+	local res = ""
+	if info.turns_left ~= nil then
+		res = res .. "<b>" .. _("Turns finished early: ") .. info.turns_left .. "</b>\n"
+	end
+	if info.finishing_bonus_per_turn ~= nil then
+		res = res .. "<small>" .. _("Early finish bonus: ") .. info.finishing_bonus_per_turn .. _(" per turn") .. "</small>\n"
+	end
+	res = res .. "<small>" .. _("Total bonus: ") .. info.finishing_bonus .. "</small>"
+
+	return res
+end
+
+---@param side side
+---@param info table
```

Comment:
Please specify the format of this table. Also, add a comment explaining what the function does.